core/arch/x86_64/mm/core_mmu.c

```diff
@@ -1222,450 +1261,1181 @@ void __weak core_init_mmu_map(unsigned long seed, struct core_mmu_config *cfg)
+	*updated_mmu_flags = current_entry;
 
-	idx = core_mmu_va2idx(tbl_info, region->va);
-	end = core_mmu_va2idx(tbl_info, region->va + region->size);
-	pa = region->pa;
+	if (!(current_entry & X86_MMU_PG_RW) && (new_entry & X86_MMU_PG_RW)) {
+		*updated_mmu_flags |= X86_MMU_PG_RW;
+		ret = true;
+	}
 
-	while (idx < end) {
-		core_mmu_set_entry(tbl_info, idx, pa, region->attr);
-		idx++;
-		pa += 1 << tbl_info->shift;
+	if (!(current_entry & X86_MMU_PG_U) && (new_entry & X86_MMU_PG_U)) {
+		*updated_mmu_flags |= X86_MMU_PG_U;
+		ret = true;
+	}
+
+	if ((current_entry & X86_MMU_PG_NX) && !(new_entry & X86_MMU_PG_NX)) {
+		*updated_mmu_flags &= ~X86_MMU_PG_NX;
+		ret = true;
+	}
+
+	/* If current entry has cached disabled and
+	 * new entry wants enable caches
+	 * let's enable them for page directory entries
+	 */
+	if ((current_entry & (X86_MMU_PG_PCD | X86_MMU_PG_PWT)) &&
+		!(new_entry & (X86_MMU_PG_PCD | X86_MMU_PG_PWT))) {
+		*updated_mmu_flags &= ~(X86_MMU_PG_PCD | X86_MMU_PG_PWT);
+		ret = true;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief  check if the physical address is valid and aligned
+ *
+ */
+static bool x86_mmu_check_paddr(paddr_t paddr)
+{
+	uint64_t addr = (uint64_t)paddr;
+	uint64_t max_paddr;
+
+    /* Check to see if the address is PAGE aligned */
+	if (!IS_ALIGNED(addr, PAGE_SIZE))
+		return false;
+
+	max_paddr = ((uint64_t)1ull << g_paddr_width) - 1;
+
+	return addr <= max_paddr;
+}
+
+/**
+ * @brief  check if the virtual address is aligned and canonical
+ *
+ */
+static bool x86_mmu_check_vaddr(vaddr_t vaddr)
+{
+	uint64_t addr = (uint64_t)vaddr;
+	uint64_t max_vaddr_lohalf, min_vaddr_hihalf;
+
+    /* Check to see if the address is PAGE aligned */
+	if (!IS_ALIGNED(addr, PAGE_SIZE))
+		return false;
+
+    /* get max address in lower-half canonical addr space */
+    /* e.g. if width is 48, then 0x00007FFF_FFFFFFFF */
+	max_vaddr_lohalf = ((uint64_t)1ull << (g_vaddr_width - 1)) - 1;
+
+    /* get min address in higher-half canonical addr space */
+    /* e.g. if width is 48, then 0xFFFF8000_00000000*/
+	min_vaddr_hihalf = ~max_vaddr_lohalf;
+
+    /* Check to see if the address in a canonical address */
+	if ((addr > max_vaddr_lohalf) && (addr < min_vaddr_hihalf))
+		return false;
+
+	return true;
+}
+
+/**
+ * @brief  Walk the page table structures
+ *
+ * In this scenario,
+ * we are considering the paging scheme to be a PAE mode with
+ * 4KB pages.
+ */
+static int x86_mmu_get_mapping(map_addr_t pml4, vaddr_t vaddr, uint32_t *ret_level,
+		arch_flags_t *mmu_flags, map_addr_t *last_valid_entry)
+{
+	uint64_t pml4e, pdpe, pde, pte;
+
+	assert(pml4);
+	if ((!ret_level) || (!last_valid_entry) || (!mmu_flags))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	*ret_level = PML4_L;
+	*last_valid_entry = pml4;
+	*mmu_flags = 0;
+
+	pml4e = get_pml4_entry_from_pml4_table(vaddr, pml4);
+	if ((pml4e & X86_MMU_PG_P) == 0)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	//  FMSG("pml4 @ 0x%lx = pml4e value 0x%lx\n", pml4, pml4e);
```

Comment:
remove the log?

Reply:
Suggest remain the log until debug is finished.